main/src/ca/uwaterloo/flix/language/ast/Scheme.scala

```diff
@@ -137,14 +138,24 @@ object Scheme {
     val (tconstrs2, tpe2) = instantiate(sc2, InstantiateMode.Rigid)
 
     // Attempt to unify the two instantiated types.
-    Unification.unifyTypes(tpe1, tpe2) match {
-      case Result.Ok(subst) =>
-        val newTconstrs1 = tconstrs1.map(subst.apply)
-        val newTconstrs2 = tconstrs2.map(subst.apply)
-        // type constraints on tvars in `sc1` must be a subset of those in `sc2`
-        // we ignore type constraints on non-vars for the purposes of the leq check
-        newTconstrs1.filter(_.arg.isInstanceOf[Type.Var]).forall(newTconstrs2.contains)
-      case Result.Err(_) => false
+    val result = for {
+      subst <- Unification.unifyTypes(tpe1, tpe2)
+      newTconstrs1 = tconstrs1.map(subst.apply)
+      newTconstrs2 = tconstrs2.map(subst.apply)
+      splitTconstrs <- ContextReduction.split(instances, Nil, Nil, newTconstrs1) // MATT not using this split; can remove?
+      (deferredTconstrs, retainedTconstrs) = splitTconstrs
+    } yield retainedTconstrs.forall(ContextReduction.entail(instances, newTconstrs2, _)) && deferredTconstrs.isEmpty
+    // deferredTconstrs contains the constraints that need to be validated by the surrounding scope
+    // This must be empty because we injected the constraints from the surrounding scope into the instance defs
+
+    // MATT probably need to split up the leq logic, return a monad
+
+    // MATT add docs
+    // MATT this checks for instances as well. does it belong here? If so, need better errors sent back to caller
+    // MATT clean up
+    result match {
```

Comment:
Can you point me to the paper you are following here?

Reply:
I'm using Typing Haskell in Haskell. The `split` function used here is in section 11.5. Do you have any particular concerns?